### ISSUE TITLE
`MultiMarginLoss` Trigger out-of-bound Read under Compute Sanitizer

### ISSUE THREAD

**Issue by Kristoff-starling** [pytorch/pytorch#88724]
### 🐛 Describe the bug

A test case for `torch.nn.MultiMarginLoss` triggers out-of-bound read error with compute sanitizer.
Test:
```python
import torch

def test():
    arg_class = torch.nn.MultiMarginLoss()
    arg_1_0_tensor = torch.rand([3, 5], dtype=torch.float32)
    arg_1_0 = arg_1_0_tensor.clone().cuda()
    arg_1_1_tensor = torch.randint(-1,0,[3], dtype=torch.int64)
    arg_1_1 = arg_1_1_tensor.clone().cuda()
    arg_1 = [arg_1_0,arg_1_1,]
    res = arg_class(*arg_1)

test()
```
Error log:
```
========= COMPUTE-SANITIZER
========= Invalid __global__ read of size 4 bytes
=========     at 0x1a0 in void at::native::<unnamed>::MultiMarginLoss_forward_kernel<(int)1, float>(T2 *, T2 *, long *, T2 *, int, int, bool, T2)
=========     by thread (0,0,0) in block (0,0,0)
=========     Address 0x7f7688fffffc is out of bounds
=========     and is 4 bytes before the nearest allocation at 0x7f7689000000 of size 2,097,152 bytes
========= 
========= Invalid __global__ read of size 4 bytes
=========     at 0x1a0 in void at::native::<unnamed>::MultiMarginLoss_forward_kernel<(int)1, float>(T2 *, T2 *, long *, T2 *, int, int, bool, T2)
=========     by thread (1,0,0) in block (0,0,0)
=========     Address 0x7f7688fffffc is out of bounds
=========     and is 4 bytes before the nearest allocation at 0x7f7689000000 of size 2,097,152 bytes
========= 
========= Invalid __global__ read of size 4 bytes
=========     at 0x1a0 in void at::native::<unnamed>::MultiMarginLoss_forward_kernel<(int)1, float>(T2 *, T2 *, long *, T2 *, int, int, bool, T2)
=========     by thread (2,0,0) in block (0,0,0)
=========     Address 0x7f7688fffffc is out of bounds
=========     and is 4 bytes before the nearest allocation at 0x7f7689000000 of size 2,097,152 bytes
========= 
========= Invalid __global__ read of size 4 bytes
=========     at 0x1a0 in void at::native::<unnamed>::MultiMarginLoss_forward_kernel<(int)1, float>(T2 *, T2 *, long *, T2 *, int, int, bool, T2)
=========     by thread (3,0,0) in block (0,0,0)
=========     Address 0x7f7688fffffc is out of bounds
=========     and is 4 bytes before the nearest allocation at 0x7f7689000000 of size 2,097,152 bytes
========= 
========= Invalid __global__ read of size 4 bytes
=========     at 0x1a0 in void at::native::<unnamed>::MultiMarginLoss_forward_kernel<(int)1, float>(T2 *, T2 *, long *, T2 *, int, int, bool, T2)
=========     by thread (4,0,0) in block (0,0,0)
=========     Address 0x7f7688fffffc is out of bounds
=========     and is 4 bytes before the nearest allocation at 0x7f7689000000 of size 2,097,152 bytes
========= 
========= ERROR SUMMARY: 5 errors
```

### Versions

```
PyTorch version: 1.14.0a0+git81042d3
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 11.1.0-6
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.6 (main, Nov  2 2022, 18:53:38) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-52-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.124
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090

Nvidia driver version: 515.65.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.4.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.4.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.5
[pip3] pytorch-san==1.14.0a0+git81042d3
[pip3] torch==1.14.0a0+git81042d3
[pip3] torchvision==0.13.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] numpy                     1.21.5           py39he7a7128_1  
[conda] numpy-base                1.21.5           py39hf524024_1  
[conda] numpydoc                  1.2                pyhd3eb1b0_0
```

cc @ezyang @gchanan @zou3519 @ngimel

**Comment by albanD** [comment]
cc @ptrblck to take a closer look at this.

**Comment by malfet** [comment]
Recompiling `MultiMarginLoss.cu` with `-G` flag shows the line number:
```
/usr/local/cuda-11.6/bin/compute-sanitizer  python -c "import torch;arg_class = torch.nn.MultiMarginLoss();print(arg_class(torch.rand([3, 5], dtype=torch.float32, device='cuda'), torch.randint(-1,0,[3], dtype=torch.int64, device='cuda')))"
========= Invalid __global__ read of size 4 bytes
=========   at 0x830 in /home/nshulga/git/pytorch/pytorch/aten/src/ATen/native/cuda/MultiMarginLoss.cu:34:void at::native::<unnamed>::MultiMarginLoss_forward_kernel<(int)1, float>(T2 *, T2 *, long *, T2 *, int, int, bool, T2)
```

which points to the following line in the code:
https://github.com/pytorch/pytorch/blob/b42cc19c88675eeb4de706f394e9d0579454740f/aten/src/ATen/native/cuda/MultiMarginLoss.cu#L22

And according to the doc target values should be in range `[0,dim-1]`
